numba/cuda/tests/cudapy/test_sync.py

```diff
@@ -71,6 +90,30 @@ def test_dyn_shared_memory(self):
         compiled[1, shape, 0, ary.size * 4](ary)
         self.assertTrue(np.all(ary == 2 * np.arange(ary.size, dtype=np.int32)))
 
+    def test_threadfence_codegen(self):
+        compiled = cuda.jit("void(int32[:])")(use_threadfence)
+        ary = np.zeros(10, dtype=np.int32)
+        compiled[1, 1](ary)
```

Comment:
Perhaps a test where each thread writes its thread ID to shared memory, and then each thread verifies another thread's written value? For example in the case where warp-level fences are used, launching a kernel with 1 warp where each thread writes `smem[tid] = tid`, then checks that `smem[check_tid] == check_tid` where `check_tid = 32 - tid` - there every thread is reading another thread's written value. If it's hard to make this fail in practice, we could attempt to add shared memory bank conflicts with a strided access to shared memory, in the hope that this would serialise the writes and make it more likely that threads would not see other threads' writes if the thread fence were not working.